src/cryptonote_protocol/peerinfo_manager.h

```diff
@@ -0,0 +1,413 @@
+
+    struct hash {
+      std::size_t operator()(const peer_info &p) const {
+        // We don't care about the bool value, so we only hash the UUID
+        return boost::hash<boost::uuids::uuid>()(p.get_connection_id());
+      }
+    };
+
+    peer_info &operator=(const peer_info &other) = delete;
+    peer_info(const peer_info &other) = delete;
+    bool operator>(const peer_info &other) const = delete;
+    bool operator<(const peer_info &other) const = delete;
+
+    bool operator==(const peer_info &other) const {
+      epee::read_lock r_lock(m_data->m_mutex);
+      return m_data->connection_id == other.get_connection_id();
+    }
+
+    bool operator!=(const peer_info &other) const {
+      epee::read_lock r_lock(m_data->m_mutex);
+      return m_data->connection_id != other.get_connection_id();
+    }
+
+    void reset() {
+      epee::write_lock w_lock(m_data->m_mutex);
+      m_data->announcements.clear();
+      m_data->received = 0;
+      m_data->requested_from_me = 0;
+      m_data->sent = 0;
+    }
+
+    void add_announcement(const crypto::hash &tx_hash) {
+      epee::write_lock w_lock(m_data->m_mutex);
+      m_data->announcements.insert(tx_hash);
+    }
+
+    void add_received() {
+      epee::write_lock w_lock(m_data->m_mutex);
+      ++m_data->received;
+    }
+
+    void add_requested_from_me() {
+      epee::write_lock w_lock(m_data->m_mutex);
+      ++m_data->requested_from_me;
+    }
+
+    void add_requested_from_peer() {
+      epee::write_lock w_lock(m_data->m_mutex);
+      ++m_data->requested_from_peer;
+    }
+
+    void add_sent() {
+      epee::write_lock w_lock(m_data->m_mutex);
+      ++m_data->sent;
+    }
+
+    void add_missed() {
+      epee::write_lock w_lock(m_data->m_mutex);
+      ++m_data->missed;
+    }
+
+    size_t get_announcement_size() const {
+      epee::read_lock r_lock(m_data->m_mutex);
+      return m_data->announcements.size();
+    }
+
+    size_t get_received() const {
+      epee::read_lock r_lock(m_data->m_mutex);
+      return m_data->received;
+    }
+
+    size_t get_requested_from_me() const {
+      epee::read_lock r_lock(m_data->m_mutex);
+      return m_data->requested_from_me;
+    }
+
+    size_t get_requested_from_peer() const {
+      epee::read_lock r_lock(m_data->m_mutex);
+      return m_data->requested_from_peer;
+    }
+
+    size_t get_sent() const {
+      epee::read_lock r_lock(m_data->m_mutex);
+      return m_data->sent;
+    }
+
+    const boost::uuids::uuid &get_connection_id() const {
+      return m_data->connection_id;
+    }
+
+    size_t get_total() const {
+      epee::read_lock r_lock(m_data->m_mutex);
+      return m_data->announcements.size() + m_data->received + m_data->requested_from_me +
+             m_data->requested_from_peer + m_data->sent;
+    }
+
+    size_t get_missed() const {
+      epee::read_lock r_lock(m_data->m_mutex);
+      return m_data->missed;
+    }
+
+    std::string get_info() const {
```

Comment:
Just make this `std::ostream operator<<(...)`. Its more efficient.

Reply:
I got a bit lazy and didn't see if it was being used outside of `std::ostream` cases. So safely ignore if not always in use.